### PR TITLE
Integrate wasmd

### PR DESCRIPTION
**Pull request by KonradStaniec**
Initial integration with wasmd.

Currently it uses this commit of wasmd repo - https://github.com/CosmWasm/wasmd/tree/7e936c7fffb6f489ed9ecb797a7a2823a032b10b

There will be at least one follow up task - https://github.com/babylonchain/babylon/issues/335

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -564,6 +620,44 @@ func NewBabylonApp(
...
+		authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+		wasmOpts...,
+	)
+
+	// The gov proposal types can be individually enabled
+	if len(wasmEnabledProposals) != 0 {
+		govRouter.AddRoute(wasm.RouterKey, wasm.NewWasmProposalHandler(app.WasmKeeper, wasmEnabledProposals))
+	}
+
+	// Set legacy router for backwards compatibility with gov v1beta1
```
Comment (vitsalis) [inline app/app.go:658]: Do we need such backwards compatibility given that we're already using the latest cosmos sdk version?

File: Makefile [hunk 2]
```diff
@@ -92,6 +92,10 @@ ifeq (boltdb,$(findstring boltdb,$(BABYLON_BUILD_OPTIONS)))
   ldflags += -X github.com/cosmos/cosmos-sdk/types.DBBackend=boltdb
 endif
 
+ifeq ($(LINK_STATICALLY),true)
+	ldflags += -linkmode=external -extldflags "-Wl,-z,muldefs -static"
```
Comment (vitsalis) [inline Makefile:96]: Why do we need those extra flags?

File: contrib/images/babylond-dlv/Dockerfile [hunk 3]
```diff
@@ -4,8 +4,19 @@ RUN go install github.com/go-delve/delve/cmd/dlv@latest
...
+    wget https://github.com/CosmWasm/wasmvm/releases/download/$WASMVM_VERSION/libwasmvm_muslc.$(uname -m).a \
+        -O /lib/libwasmvm_muslc.a && \
+    # verify checksum
+    wget https://github.com/CosmWasm/wasmvm/releases/download/$WASMVM_VERSION/checksums.txt -O /tmp/checksums.txt && \
+    sha256sum /lib/libwasmvm_muslc.a | grep $(cat /tmp/checksums.txt | grep $(uname -m) | cut -d ' ' -f 1)
+
+# Copy the remaining files
+COPY . .
+
+RUN LEDGER_ENABLED=false BUILD_TAGS=muslc LINK_STATICALLY=true make COSMOS_BUILD_OPTIONS="debug,nostrip" clean build
```
Comment (vitsalis) [inline contrib/images/babylond-dlv/Dockerfile:19]: Why do we add `BUILD_TAGS=muslc`?

**Review by vitsalis — COMMENTED** [review]
Mostly a few questions:

File: app/app.go [hunk 4]
```diff
@@ -161,22 +166,25 @@ var (
 		mint.AppModuleBasic{},
 		distr.AppModuleBasic{},
 		gov.NewAppModuleBasic(
-			[]govclient.ProposalHandler{
+			append(
+				wasmclient.ProposalHandlers,
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply app/app.go:170]: tbh mostly copied from wasmd repo, although from what I understand code the order does not really matter here, it only enables some cli commands.

File: app/app.go [hunk 5]
```diff
@@ -161,22 +166,25 @@ var (
...
 				paramsclient.ProposalHandler,
 				upgradeclient.LegacyProposalHandler,
 				upgradeclient.LegacyCancelProposalHandler,
 				ibcclientclient.UpdateClientProposalHandler,
 				ibcclientclient.UpgradeProposalHandler,
-			},
+			),
 		),
 		params.AppModuleBasic{},
+		consensus.AppModuleBasic{},
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply app/app.go:179]: it enables us to change consensus params through government proposal, seems useful :D

File: app/app.go [hunk 6]
```diff
@@ -206,6 +214,46 @@ var (
...
+	// EnableSpecificWasmProposals, if set, must be comma-separated list of values
+	// that are all a subset of "EnableAllProposals", which takes precedence over
+	// WasmProposalsEnabled.
+	//
+	// See: https://github.com/CosmWasm/wasmd/blob/02a54d33ff2c064f3539ae12d75d027d9c665f05/x/wasm/internal/types/proposal.go#L28-L34
+	EnableSpecificWasmProposals = ""
+
+	// EmptyWasmOpts defines a type alias for a list of wasm options.
+	EmptyWasmOpts []wasm.Option
+)
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply app/app.go:233]: unfortunately you cannot have `const` slice in golang

File: app/app.go [hunk 7]
```diff
@@ -564,6 +620,44 @@ func NewBabylonApp(
...
+	wasmDir := filepath.Join(homePath, "wasm")
+	wasmConfig, err := wasm.ReadWasmConfig(appOpts)
+	if err != nil {
+		panic(fmt.Sprintf("error while reading wasm config: %s", err))
+	}
+
+	// The last arguments can contain custom message handlers, and custom query handlers,
+	// if we want to allow any custom callbacks
+	// See https://github.com/CosmWasm/cosmwasm/blob/main/docs/CAPABILITIES-BUILT-IN.md
+	availableCapabilities := "iterator,staking,stargate,cosmwasm_1_1,cosmwasm_1_2"
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply app/app.go:632]: `iterator` seems like really basic feature to have (it enables iteration over keystore in smart contract)
`stargate` is IBC so seems useful
`cosmwasm_1_1,cosmwasm_1_2` are just some new feature added to cosmwasm, so it is probably be good to be up to date

One thing which is debatable is `staking`.

In general, I can create some ADR which describes all capabilities first and then we may talk it out on some eng meeting. Current version is just reasonable default. I will left TODO here.

File: app/app.go (hunk 1 above)
Comment (KonradStaniec) [reply app/app.go:658]: tbh every project and example app does it (in cosmos-sdk, ibc and wasmd rep), so maybe there is some tooling which support those legacy proposals which not migrated yet. I would keep it for now and possibly review this once again later.

File: app/app.go [hunk 8]
```diff
@@ -960,5 +1090,8 @@ func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino
 	paramsKeeper.Subspace(ibctransfertypes.ModuleName)
 	paramsKeeper.Subspace(zctypes.ModuleName)
 
+	// Wasm
+	paramsKeeper.Subspace(wasm.ModuleName)
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply app/app.go:1094]: hmm we can remove it from here as wasm keeper keeps its owns params it seems.

File: Makefile (hunk 2 above)
Comment (KonradStaniec) [reply Makefile:96]: it is needed to build docker container with comoswasm on alpine.

File: contrib/images/babylond-dlv/Dockerfile (hunk 3 above)
Comment (KonradStaniec) [reply contrib/images/babylond-dlv/Dockerfile:19]: it just informational. To statically link babylon with cosmowasm on apline you need muslc, not glibc.

File: app/app.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vitsalis) [reply app/app.go]: Sounds good to me, let's nail down the capabilities we want to have enabled

**Review by vitsalis — APPROVED** [review]

**Pull request merged**